<a href="https://colab.research.google.com/github/drwitt/NLP_IDS_690-03/blob/master/Copy_of_pbloem_Transformer_test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This is a test of "TRANSFORMERS FROM SCRATCH" by P.Bloem 
www.peterbloem.nl/blog/transformers 

Also at 
https://github.com/pbloem/former

perhaps the clearest expostion of transformers


## Make sure to run it with  GPU 

**CPU**


 epoch 0
100% 5000/5000 [1:03:05<00:00,  1.45it/s]
-- validation accuracy 0.725

 epoch 1
100% 5000/5000 [1:07:57<00:00,  1.22it/s]
-- validation accuracy 0.765

**GPU** 

 epoch 0
100% 5000/5000 [05:10<00:00, 16.11it/s]
-- validation accuracy 0.7

 epoch 1
100% 5000/5000 [05:09<00:00, 14.03it/s] 
-- validation accuracy 0.755


**TPU**
 epoch 0
100% 5000/5000 [1:06:07<00:00,  1.13it/s]
-- validation accuracy 0.693





In [0]:
# Clone the entire repo.
!git clone -l -s git://github.com/pbloem/former.git former
!cd former/
!ls

Cloning into 'former'...
remote: Enumerating objects: 13, done.
remote: Counting objects: 100% (13/13), done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 108 (delta 3), reused 8 (delta 3), pack-reused 95
Receiving objects: 100% (108/108), 34.89 MiB | 8.87 MiB/s, done.
Resolving deltas: 100% (39/39), done.
former	sample_data


In [0]:
!pip install torch tb-nightly tqdm numpy torchtext

     |████████████████████████████████| 3.8MB 2.8MB/s 


In [0]:
!cd former/former
!ls 


former	sample_data


In [0]:
!ls former/experiments
!python ./former/experiments/classify.py -e 20

classify.py  _context.py  generate.py
OPTIONS  Namespace(batch_size=4, depth=6, embedding_size=128, final=False, gradient_clipping=1.0, lr=0.0001, lr_warmup=10000, max_length=512, max_pool=False, num_epochs=20, num_heads=8, seed=1, tb_dir='./runs', vocab_size=50000)
downloading aclImdb_v1.tar.gz
aclImdb_v1.tar.gz: 100% 84.1M/84.1M [00:07<00:00, 11.8MB/s]
- nr. of training examples 5000
- nr. of validation examples 1250

 epoch 0
100% 5000/5000 [05:11<00:00, 15.90it/s]
-- validation accuracy 0.693

 epoch 1
100% 5000/5000 [05:11<00:00, 14.85it/s]
-- validation accuracy 0.757

 epoch 2
100% 5000/5000 [05:10<00:00, 14.93it/s]
-- validation accuracy 0.753

 epoch 3
100% 5000/5000 [05:09<00:00, 16.16it/s]
-- validation accuracy 0.772

 epoch 4
100% 5000/5000 [05:08<00:00, 16.21it/s]
-- validation accuracy 0.792

 epoch 5
100% 5000/5000 [05:09<00:00, 16.18it/s]
-- validation accuracy 0.824

 epoch 6
100% 5000/5000 [05:08<00:00, 15.07it/s]
-- validation accuracy 0.843

 epoch 7
100% 5000/5000

#Example run

### notice some non-monotonicity in convergence

### running it for 68 epochs didn't improve the resuls



classify.py  _context.py  generate.py
OPTIONS  Namespace(batch_size=4, depth=6, embedding_size=128, final=False, gradient_clipping=1.0, lr=0.0001, lr_warmup=10000, max_length=512, max_pool=False, num_epochs=20, num_heads=8, seed=1, tb_dir='./runs', vocab_size=50000)
downloading aclImdb_v1.tar.gz
aclImdb_v1.tar.gz: 100% 84.1M/84.1M [00:07<00:00, 11.8MB/s]
- nr. of training examples 5000
- nr. of validation examples 1250

 epoch 0
100% 5000/5000 [05:11<00:00, 15.90it/s]
-- validation accuracy 0.693

 epoch 1
100% 5000/5000 [05:11<00:00, 14.85it/s]
-- validation accuracy 0.757

 epoch 2
100% 5000/5000 [05:10<00:00, 14.93it/s]
-- validation accuracy 0.753

 epoch 3
100% 5000/5000 [05:09<00:00, 16.16it/s]
-- validation accuracy 0.772

 epoch 4
100% 5000/5000 [05:08<00:00, 16.21it/s]
-- validation accuracy 0.792

 epoch 5
100% 5000/5000 [05:09<00:00, 16.18it/s]
-- validation accuracy 0.824

 epoch 6
100% 5000/5000 [05:08<00:00, 15.07it/s]
-- validation accuracy 0.843

 epoch 7
100% 5000/5000 [05:06<00:00, 16.30it/s]
-- validation accuracy 0.839

 epoch 8
100% 5000/5000 [05:06<00:00, 15.44it/s]
-- validation accuracy 0.848

 epoch 9
100% 5000/5000 [05:06<00:00, 16.30it/s]
-- validation accuracy 0.853

 epoch 10
100% 5000/5000 [05:06<00:00, 16.43it/s]
-- validation accuracy 0.853

 epoch 11
100% 5000/5000 [05:03<00:00, 15.47it/s]
-- validation accuracy 0.853

 epoch 12
100% 5000/5000 [05:03<00:00, 16.23it/s]
-- validation accuracy 0.863

 epoch 13
100% 5000/5000 [05:03<00:00, 15.07it/s]
-- validation accuracy 0.862

 epoch 14
100% 5000/5000 [05:08<00:00, 16.22it/s]
-- validation accuracy 0.852

 epoch 15
100% 5000/5000 [05:06<00:00, 16.31it/s]
-- validation accuracy 0.865

 epoch 16
100% 5000/5000 [05:07<00:00, 16.28it/s]
-- validation accuracy 0.862

 epoch 17
100% 5000/5000 [05:04<00:00, 16.41it/s]
-- validation accuracy 0.85

 epoch 18
100% 5000/5000 [05:05<00:00, 16.49it/s]
-- validation accuracy 0.86

 epoch 19
100% 5000/5000 [05:05<00:00, 15.54it/s]
-- validation accuracy 0.858